reactor-core/src/main/java/reactor/core/publisher/MonoSingleOptional.java

```diff
@@ -96,24 +93,12 @@ public void onSubscribe(Subscription s) {
 
 		@Override
 		public void onNext(T t) {
-			if (isCancelled()) {
-				//this helps differentiating a duplicate malformed signal "done" from a count > 1 "done"
-				Operators.onDiscard(t, actual().currentContext());
-				return;
-			}
 			if (done) {
 				Operators.onNextDropped(t, actual().currentContext());
 				return;
 			}
-			if (++count > 1) {
-				Operators.onDiscard(t, actual().currentContext());
-				//mark as both cancelled and done
-				cancel();
-				onError(new IndexOutOfBoundsException("Source emitted more than one item"));
-			}
-			else {
-				setValue(Optional.of(t));
-			}
+			done = true;
+			complete(Optional.of(t));
```

Comment:
I feel this change is a bit too simplistic. In `onError` there is the call to `discardTheValue()` which would discard the value in case an `onError()` followed, but now that call is redundant, as the value is not stored. Let's take a step back for a moment.
We are assuming this operator is within the reactor chain and the source is `Mono` which allows only [`onNext` followed by `onComplete`] or [`onError`] alone.
I'd suggest keeping the `setValue(Optional.of(t))` to allow for checking what `onComplete` needs to do. But still deliver the value to the subscriber in `onNext`. For the `onComplete` logic -> if the value is null -> complete with `Optional.empty()`, otherwise, just send the complete signal.
Having `onNext` and `onComplete` delivered to downstream before the upstream sends `onComplete` sounds incorrect, but I can't think of the implications. So I think it's best to keep the causal relationship and preserve the order. The previous implementation delayed delivering the value until `onComplete` from upstream happened, but that is usually correct, as the completion signal is expected to be delivered immediately after `onNext` in most cases.
For `onError` I'd skip the `discardTheValue` bit, as we're not expecting `onNext` followed by `onError` in this operator. This has its purpose if the source was a `Flux`, where discarding all improperly delivered items is vital.

Reply:
We had a discussion with @OlegDokuka and as a result of that, please disregard the above suggestions.
Specifically, Oleh clarified that it is correct to immediately complete the downstream once `onNext` arrives. The `onNext` and `onComplete` are coupled by time to optimise the call stack reuse and other operators also utilise this - e.g. `MonoSink`, `MonoHasElement`. There is no need to wait for completion from upstream, it's ok to discard it after value was received.

Having clarified the coupling, line 111 with `discardTheValue()` can be removed, as this was applicable for the `Flux` source in case of `MonoSingle.SingleSubscriber`, but is not applicable here – as explained, `onError` will not happen after `onNext` in case of `Mono`.

Reply:
Ok. Yes I already wanted to respond that MonoFlatMap is also completing in `onNext()` for example.
I'll update the code later today.

I have a more general question. When reading through other classes I don't always understand in which case plain java fields are used (like the `done` field updated in `onNext()` without any synchronization, and in which case `compareAndSet` semantic is applied to guarantee some atomic update and/or thread-safety/data visibility.

For example `MonoFlatMap.FlatMapInner.onSubscribe()` uses compareAndSet to make sure there will never be more than 1 subscription which could only happen if multiple `onNext()` call are processed in parallel. But doesn't the framework already guarantee that `onNext()` calls come in sequentially? And that for a Mono there is only 1 item at most?

My code in `MonoSingleOptional` simply updates the `done` field without any synchronization.
I realized the `complete(Optional.of(t))` code I'm calling is also using `compareAndSet`. Here I understand it better, as it is used to update a state that can be modified `onNext()`, `request(n)` or `cancel()` calls which can be in parallel on different threads.

Is there a guide somewhere explaining what thread-safety guarantees are given by the framework and which responsibility has an individual operator ? Do operators need to protect themselves against bad implementations of subscribers? Or whether they can assume the `[onNext, onNext, ... completed]` sequence comes in serially and on the same thread? And assume a mono will never issue multiple items?

I don't really need answers to these questions to finish my work. But in case you have some existing doc about these questions I'd be interested to read it.

Reply:
>For example MonoFlatMap.FlatMapInner.onSubscribe() uses compareAndSet to make sure there will never be more than 1 subscription which could only happen if multiple onNext() call are processed in parallel. But doesn't the framework already guarantee that onNext() calls come in sequentially? And that for a Mono there is only 1 item at most?

This happens in the call to replace `FlatMapMain.second` field, which is also accessed from `cancel`, which by the principle you described later, can be called from a different thread, hence non-blocking synchronisation happens.

In general the rules around thread safety are described in the [reactive-streams specification](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#specification)

>And assume a mono will never issue multiple items?

Here, when we are dealing with the source of type `Mono`, we should be able to assume it's acting according to the documentation of `Mono`. When dealing with a generic `Publisher` though, more precautions need to be taken. Some operators add more safeguards than others it seems from my experience. But in general when we're dealing with an internal operator, we can assume compliance to the spec.

Reply:
Thanks for the clear explanation!